lib/src/client.dart

```diff
@@ -955,13 +957,63 @@ class Client extends MatrixApi {
   final Map<String, ProfileInformation> _profileRoomsCache = {};
   final Map<String, ProfileInformation> _profileServerCache = {};
 
+  /// Get the combined profile information for this user. First checks for a
+  /// non outdated cached profile before requesting from the server. Cached
+  /// profiles are outdated if they have been cached in a time older than the
+  /// [maxCacheAge] or they have been marked as outdated by an event in the
+  /// sync loop.
+  /// In case of an
+  ///
+  /// [userId] The user whose profile information to get.
+  @override
+  Future<CachedProfileInformation> getUserProfile(
+    String userId, {
+    Duration timeout = const Duration(seconds: 30),
+    Duration maxCacheAge = const Duration(days: 1),
+  }) async {
+    final cachedProfile = await database?.getUserProfile(userId);
+    if (cachedProfile != null &&
+        (cachedProfile.outdated != true ||
+            DateTime.now().difference(cachedProfile.updated) >= maxCacheAge)) {
+      return cachedProfile;
+    }
+
+    late final ProfileInformation profile;
+    try {
+      profile = await (_userProfileRequests[userId] ??=
+          super.getUserProfile(userId).timeout(timeout));
+    } catch (e) {
+      Logs().d('Unable to fetch profile from server', e);
+      if (cachedProfile == null) rethrow;
+      return cachedProfile;
+    } finally {
+      unawaited(_userProfileRequests.remove(userId));
+    }
+
+    final newCachedProfile = CachedProfileInformation.fromProfile(
+      profile,
+      outdated: false,
+      updated: DateTime.now(),
+    );
+
+    await database?.storeUserProfile(userId, newCachedProfile);
+
+    return newCachedProfile;
+  }
+
+  final Map<String, Future<ProfileInformation>> _userProfileRequests = {};
+
+  final CachedStreamController<String> onUserProfileUpdate =
+      CachedStreamController<String>();
+
   /// Get the combined profile information for this user.
   /// If [getFromRooms] is true then the profile will first be searched from the
   /// room memberships. This is unstable if the given user makes use of different displaynames
   /// and avatars per room, which is common for some bots and bridges.
   /// If [cache] is true then
   /// the profile get cached for this session. Please note that then the profile may
   /// become outdated if the user changes the displayname or avatar in this session.
+  @Deprecated('User `getUserProfile(userID)` instead')
   Future<Profile> getProfileFromUserId(String userId,
```

Comment:
can we also change the usage of `getProfileFromUserId` in `Room.requestUser()` to `getUserProfile`? Please check if my suggestion below for `User` class makes sense too

Reply:
I'm talking about this line -

https://github.com/famedly/matrix-dart-sdk/blob/191397ebb8ba7b93b288365c8cedc5941289140d/lib/src/room.dart#L1720

Reply:
But this is already changed 🤔